File: java/209.minimum-size-subarray-sum.java
/*
 * @lc app=leetcode id=209 lang=java
 *
 * [209] Minimum Size Subarray Sum
 */
class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int min = nums.length + 1, sum = 0;
        int i = 0, j = -1;
        while (i < nums.length) {
            if (sum < s && j < nums.length - 1) sum += nums[++j];
            else sum -= nums[i++];
            if (sum >= s && j-i+1 < min) min = j-i+1;
        }
        return min == nums.length + 1 ? 0 : min;
    }
}

File: go/209.minimum-size-subarray-sum.go
/*
 * @lc app=leetcode id=209 lang=golang
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
func minSubArrayLen(s int, nums []int) int {
	min := len(nums) + 1
	sum := 0
	for i, j := 0, 0; i < len(nums); {
		if sum < s && j < len(nums) {
			sum += nums[j]
			j++
		} else {
			sum -= nums[i]
			i++
		}
		if sum >= s && min > j-i {
			min = j - i
		}
	}
	if min == len(nums)+1 {
		return 0
	} else {
		return min
	}
}

// @lc code=end

